archspec/cpu/detect.py

```diff
@@ -206,11 +206,26 @@ def host():
     # Get a list of possible candidates for this micro-architecture
     candidates = compatible_microarchitectures(info)
 
+    # Sorting criteria for candidates
+    def sorting_fn(item):
+        return len(item.ancestors), len(item.features)
+
+    # Get the best generic micro-architecture
+    generic_candidates = [c for c in candidates if c.vendor == "generic"]
+    best_generic = sorted(generic_candidates, key=sorting_fn, reverse=True)[0]
+
+    # Filter the candidates to be descendant of the best generic candidate.
+    # This is to avoid that the lack of a niche feature that can be disabled
+    # from e.g. BIOS prevents detection of a reasonably performant architecture
+    candidates = [c for c in candidates if c > best_generic]
+
+    # If we don't have candidates, return the best generic micro-architecture
+    if not candidates:
+        return best_generic
+
     # Reverse sort of the depth for the inheritance tree among only targets we
     # can use. This gets the newest target we satisfy.
-    return sorted(
-        candidates, key=lambda t: (len(t.ancestors), len(t.features)), reverse=True
-    )[0]
+    return sorted(candidates, key=sorting_fn, reverse=True)[0]
```

Comment:
I think using `sorted` here is less efficient than it could be. Can't you use graph traversal instead of sorting?  Asking b/c IIRC our comparison operator does a graph traversal right?  i.e., it looks to see if one thing is an ancestor of another.  That makes operations like `sorted()` end up being quadratic, because each comparison traverses the graph.

Don't we already know the sort order because we have the graph? I was thinking this algorithm could be implemented as a simple depth-first search, starting at each generic architecture and proceeding to newer ones until we hit one we're not compatible with.

Reply:
> Can't you use graph traversal instead of sorting?

We can do that. Iirc when we first implemented detection we accepted the algorithm being quadratic given that the DAG would always be of very small size. I'm fine implementing this, but can we keep it in a separate PR?

Reply:
+1 on not overoptimising this since the DAG is quite small, but if it's done it should definitely be a separate PR.